### PR TITLE
Gracefully handle documented exceptions from AssemblyMetadata.GetModules

### PR DESCRIPTION
**Pull request by mavasani**
We have done so in past for other callsites to this API, such as #17687. This change does the same for the remaining two callsites in the IDE layer.

Fixes VSO Watson [234447](https://devdiv.visualstudio.com/DevDiv/_queries/edit/234447): _CLR_EXCEPTION_System.BadImageFormatException_8007000b_Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.PEModule.get_ReferencedAssemblies_

### CODE REVIEW

File: src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs [hunk 1]
```diff
@@ -320,6 +325,23 @@ private void WriteMvidsTo(Metadata metadata, ObjectWriter writer, CancellationTo
             WriteMvidTo((ModuleMetadata)metadata, writer, cancellationToken);
         }
 
+        private static bool TryGetModules(AssemblyMetadata assemblyMetadata, out ImmutableArray<ModuleMetadata> modules)
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs]: Could we make this an extension method and share it with code in  #17687?

File: src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs [hunk 2]
```diff
@@ -304,9 +304,14 @@ private void WriteMvidsTo(Metadata metadata, ObjectWriter writer, CancellationTo
 
             if (metadata is AssemblyMetadata assemblyMetadata)
             {
+                if (!TryGetModules(assemblyMetadata, out var modules))
+                {
+                    // Gracefully bail out without writing anything to the writer.
```
Comment (genlu) [inline src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs]: Is `writer.WriteInt32(MetadataFailed);` needed here?

Comment (mavasani) [reply src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs]: I am matching the implementation for other bail out case in this method.

File: src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs (hunk 1 above)
Comment (mavasani) [reply src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs]: Yes, let me make this an extension.

Comment (mavasani) [reply src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs]: Actually, seems that callsite has a try-catch for two different method calls and also returns modules - using this extension would just make the code in there more icky. I'll stick with the current code style for now..

**Comment by mavasani** [comment]
Thanks @genlu.

@vatsalyaagrawal @jinujoseph for approval

**Pull request merged**